nrf-softdevice/src/ble/advertisement_builder.rs

```diff
@@ -138,145 +160,163 @@ pub enum BasicService {
+        }
     }
 
-    /// View the resulting advertisement data in the form of a byte slice.
-    pub fn as_slice(&self) -> Result<&[u8], Error> {
-        (self.ptr <= K)
-            .then(|| &self.buf[..self.ptr])
-            .ok_or(Error::Oversize { expected: self.ptr })
+    /// Get the resulting advertisement payload.
+    ///
+    /// Panics if more than `K` bytes were written to the builder.
+    pub const fn build(self) -> AdvertisementPayload<K> {
+        // Use core::assert! even if defmt is enabled because it is const
+        core::assert!(self.ptr <= K, "advertisement exceeded buffer length");
+
+        AdvertisementPayload {
+            buf: self.buf,
+            len: self.ptr,
+        }
     }
 
     /// Add flags to the advertisement data.
-    pub fn flags<const N: usize>(self, flags: [Flag; N]) -> Self {
-        let result = flags.iter().fold(0, |partial, &flag| partial + flag as u8);
+    pub const fn flags(self, flags: &[Flag]) -> Self {
+        let mut i = 0;
+        let mut bits = 0;
+        while i < flags.len() {
+            bits |= flags[i] as u8;
+            i += 1;
+        }
 
-        self.raw(ADType::Flags, &[result])
+        self.raw(AdvertisementDataType::FLAGS, &[bits])
     }
 
-    /// Add a list of services to the advertisement data.
-    pub fn services<L: ServiceList<S, N>, S: Service, const N: usize>(mut self, services: L) -> Self {
-        self.write(&[(N * S::SIZE) as u8 + 1, L::AD as u8]);
+    /// Add a list of 16-bit service uuids to the advertisement data.
+    pub const fn services_16(self, complete: ServiceList, services: &[BasicService]) -> Self {
+        let ad_type = match complete {
+            ServiceList::Incomplete => AdvertisementDataType::INCOMPLETE_16_SERVICE_LIST,
+            ServiceList::Complete => AdvertisementDataType::COMPLETE_16_SERVICE_LIST,
+        };
+
+        let mut res = self.write(&[(services.len() * 2) as u8 + 1, ad_type.to_u8()]);
+        let mut i = 0;
+        while i < services.len() {
+            res = res.write(&(services[i] as u16).to_le_bytes());
+            i += 1;
+        }
+        res
+    }
 
-        for service in services.list() {
-            service.render(&mut self);
+    /// Add a list of 128-bit service uuids to the advertisement data.
+    ///
+    /// Note that each UUID in the list needs to be in little-endian format, i.e. opposite to what you would
+    /// normally write UUIDs.
+    pub const fn services_128(self, complete: ServiceList, services: &[[u8; 16]]) -> Self {
+        let ad_type = match complete {
+            ServiceList::Incomplete => AdvertisementDataType::INCOMPLETE_128_SERVICE_LIST,
+            ServiceList::Complete => AdvertisementDataType::COMPLETE_128_SERVICE_LIST,
+        };
+
+        let mut res = self.write(&[(services.len() * 16) as u8 + 1, ad_type.to_u8()]);
+        let mut i = 0;
+        while i < services.len() {
+            res = res.write(&services[i]);
+            i += 1;
         }
```

Comment:
The splitting of these functions is really unfortunate, but seemingly unavoidable. Looking forward to const traits.